pkg/policy/k8s/service.go

```diff
@@ -0,0 +1,278 @@
+
+// onServiceEvent processes a ServiceNotification and (if necessary)
+// recalculates all policies affected by this change.
+func (p *PolicyWatcher) onServiceEvent(event k8s.ServiceNotification) {
+	err := p.updateToServicesPolicies(event.ID, event.Service, event.OldService)
+	if err != nil {
+		p.log.WithError(err).WithFields(logrus.Fields{
+			logfields.Event:     event.Action,
+			logfields.ServiceID: event.ID,
+		}).Warning("Failed to recalculate CiliumNetworkPolicy rules after service event")
+	}
+}
+
+// updateToServicesPolicies is to be invoked when a service has changed (i.e. it was
+// added, removed, its endpoints have changed, or its labels have changed).
+// This function then checks if any of the known CNP/CCNPs are affected by this
+// change, and recomputes them by calling resolveCiliumNetworkPolicyRefs.
+func (p *PolicyWatcher) updateToServicesPolicies(svcID k8s.ServiceID, newSVC, oldSVC *k8s.Service) error {
+	var errs []error
+
+	// Bail out early if updated service is not selectable
+	if !(p.isSelectableService(newSVC) || p.isSelectableService(oldSVC)) {
+		return nil
+	}
+
+	// If the service selector or its labels have changed,
+	// we have to also want to update any CNPs that matched oldSVC
+	checkOldService := oldSVC != nil && !newSVC.DeepEqual(oldSVC)
+
+	for key, cnp := range p.cnpCache {
+		if !(p.cnpMatchesService(cnp, svcID, newSVC) ||
+			(checkOldService && p.cnpMatchesService(cnp, svcID, oldSVC))) {
+			continue
+		}
+
+		p.log.WithFields(logrus.Fields{
+			logfields.CiliumNetworkPolicyName: cnp.Name,
+			logfields.K8sAPIVersion:           cnp.APIVersion,
+			logfields.K8sNamespace:            cnp.Namespace,
+			logfields.ServiceID:               svcID,
+		}).Debug("Service updated or deleted, recalculating CiliumNetworkPolicy rules")
+
+		initialRecvTime := time.Now()
+
+		resourceKind := ipcacheTypes.ResourceKindCNP
+		if len(key.Namespace) == 0 {
+			resourceKind = ipcacheTypes.ResourceKindCCNP
+		}
+		resourceID := ipcacheTypes.NewResourceID(
+			resourceKind,
+			cnp.ObjectMeta.Namespace,
+			cnp.ObjectMeta.Name,
+		)
+
+		errs = append(errs, p.resolveCiliumNetworkPolicyRefs(cnp, key, initialRecvTime, resourceID))
+	}
+	return errors.Join(errs...)
+}
+
+// resolveToServices translates all ToServices rules found in the provided CNP
+// and to corresponding ToCIDRSet rules. Mutates the passed in cnp in place.
+func (p *PolicyWatcher) resolveToServices(cnp *types.SlimCNP) {
+	// The svcEndpoints internally associates each ToService selector with
+	// the list of matching service endpoints.
+	svcEndpoints := emptyMatchedEndpoints()
+
+	// We consult the service cache to obtain the service endpoints
+	// which are selected by the ToServices selectors found in the CNP.
+	// The selected endpoints are stored in svcEndpoints.
+	p.svcCache.ForEachService(func(svcID k8s.ServiceID, svc *k8s.Service, eps *k8s.Endpoints) bool {
+		if p.isSelectableService(svc) {
+			svcEndpoints.extractMatches(cnp.Spec, svcID, svc, eps)
```

Comment:
why are there separate `extractMatches(spec)` and `translateRule(spec)` calls?

Reply:
That's a fair question. Mainly because `extractMatches` is invoked for every service **and** rule combination, in service random order. `translateRule` is only invoked for every rule once. Having `translate` be a separate step made it easier to have deterministic rule translation and endpoint deduplication.

But I can take a look again at this though. The overall flow has changed a bit since I first came up with the "extract, then translate" logic and maybe having a single method is much simpler now.

Reply:
Yes